diff --git a/src/system.h b/src/system.h
index 7bc778b..50777f8 100644
--- a/src/system.h
+++ b/src/system.h
@@ -29,6 +29,12 @@
 #include <stat-size.h>
 #include <stat-time.h>
 #include <timespec.h>
+#ifdef _WIN32
+#ifndef _TIMESPEC_TYPEDEF
+typedef struct timespec timespec;
+#define _TIMESPEC_TYPEDEF
+#endif
+#endif
 
 #include <unistd.h>
 
diff --git a/src/util.c b/src/util.c
index 6c65991..e14e260 100644
--- a/src/util.c
+++ b/src/util.c
@@ -283,7 +283,9 @@ process_signals (void)
       if (stops)
         {
           stop_signal_count = stops - 1;
+#ifndef _WIN32
           sig = SIGSTOP;
+#endif
         }
       else
 	{
